closed/adds/jdk/src/share/native/jdk/crypto/jniprovider/NativeCrypto.c

```diff
@@ -318,16 +353,107 @@ JNIEXPORT jint JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_loadCrypto
         (NULL == OSSL_RSA_private_encrypt) ||
         (NULL == OSSL_BN_bin2bn) ||
         (NULL == OSSL_BN_set_negative) ||
-        (NULL == OSSL_BN_free)) {
-        //fprintf(stderr, "One or more of the required symbols are missing in the crypto library\n");
-        //fflush(stderr);
+        (NULL == OSSL_BN_free) ||
+        ((NULL == OSSL_CRYPTO_num_locks) && (0 == ossl_ver)) ||
+        ((NULL == OSSL_CRYPTO_THREADID_set_numeric) && (0 == ossl_ver)) ||
+        ((NULL == OSSL_OPENSSL_malloc) && (0 == ossl_ver)) ||
+        ((NULL == OSSL_CRYPTO_THREADID_set_callback) && (0 == ossl_ver)) ||
+        ((NULL == OSSL_CRYPTO_set_locking_callback) && (0 == ossl_ver))) {
+        // fprintf(stderr, "One or more of the required symbols are missing in the crypto library\n");
+        // fflush(stderr);
         unload_crypto_library(handle);
         return -1;
     } else {
+        if (0 == ossl_ver) {
+            if (0 != thread_setup()) {
+                unload_crypto_library(handle);
+                return -1;
+            }
+        }
         return 0;
     }
  }
 
+#ifdef WINDOWS
+static HANDLE *lock_cs;
+
+int thread_setup()
+{
+    int i;
+    int size = (*OSSL_CRYPTO_num_locks)() * sizeof(HANDLE);
+    lock_cs = (*OSSL_OPENSSL_malloc)(size);
+    if (NULL == lock_cs) {
+        return -1;
+    }
+    for (i = 0; i < (*OSSL_CRYPTO_num_locks)(); i++) {
+        lock_cs[i] = CreateMutex(NULL, FALSE, NULL);
+        if (lock_cs[i] == NULL) {
+            fprintf(stderr, "CreateMutex error: %d\n", GetLastError());
+        }
+    }
+    (*OSSL_CRYPTO_set_locking_callback)((void (*)(int, int, char *, int))
+                                win32_locking_callback);
+    return 0;
+}
+
+void win32_locking_callback(int mode, int type, const char *file, int line)
+{
+    if (0 != (mode & CRYPTO_LOCK)) {
+        DWORD dwWaitResult = WaitForSingleObject(lock_cs[type], INFINITE);
+        if (dwWaitResult == WAIT_FAILED) {
+            fprintf(stderr, "WaitForSingleObject error: %d\n", GetLastError());
+        }
+    } else {
+        int releaseResult = ReleaseMutex(lock_cs[type]);
+        if (releaseResult == 0) {
+            fprintf(stderr, "ReleaseMutex error: %d\n", GetLastError());
+        }
+    }
+}
+
+#else
+static pthread_mutex_t *lock_cs;
+
+int thread_setup()
+{
+    int i;
+    int size = (*OSSL_CRYPTO_num_locks)() * sizeof(pthread_mutex_t);
+    lock_cs = (*OSSL_OPENSSL_malloc)(size);
+    if (NULL == lock_cs) {
+        return -1;
+    }
+    for (i = 0; i < (*OSSL_CRYPTO_num_locks)(); i++) {
+        int initResult = pthread_mutex_init(&(lock_cs[i]), NULL);
+        if (initResult != 0) {
+            fprintf(stderr, "pthread_mutex_init error\n");
```

Comment:
Constant on the left of `!=`.
Like Windows, this whole function should fail if we can't create enough mutexes.